### PR TITLE
Fix unloading `unloadable` constants which were autoloaded 

### LINKED ISSUE #32359: When a `unloadable` module is autoloaded, it can't be reloaded after unloading once
### Steps to reproduce

I've pushed a failing test here: https://github.com/rails/rails/commit/5be9bd575907394d8b3b47c2e7745c118d3ba495

The problem is: 

- When `unloadable` constants are autoloaded, their names are added to two lists: 
  - `Dependencies.autoloaded_constants`, which holds the names of all constants which were autoloaded; AND 
  - `Dependencies.explicitly_unloadable_constants`, which holds the names of all loaded constants which declared `unloadable` 
- When the autoloaded constant has an `unloadable` child constant, the arrays look like: 
  - `autoloaded_constants # => ["UnloadableExample"]`
  - `explicitly_unloadable_constants # => ["UnloadableExample"]`
- During `remove_unloadable_constants!`: 
  - First, `autoloaded_constants` are unloaded, so `UnloadableExample` is removed from Ruby and removed from `loaded` and `autoloaded_constants`. 
  - Then, `explicitly_unloadable_constants` are unloaded. 
    - `Object::UnloadableExample` is on the list, and `remove_constant`'s call to `parent.const_get(to_remove)` causes `UnloadableExample` to be re-autoloaded 
    - This adds its name to `autoloaded_constants` and its path to `loaded` 
    - Then, the constant is removed from Ruby 
  - Then `remove_unloadable_constants!` is done 

At this point, the state is inconsistent: 

- `autoloaded_constants` contains `["UnloadableExample"]`, since it was autoloaded during `remove_constant` 
- Ruby `defined?(UnloadableExample)` returns nil, because the constant was _actually_ removed during `remove_constant 

This _matters_ because on the next page load, you get: 

> LoadError: Unable to autoload constant UnloadableExample, expected /Users/rmosolgo/code/rails/activesupport/test/autoloading_fixtures/explicit_unloadable/unloadable_example.rb to define it

This is because `loaded` contains its path, so AS::D _thinks_ it was autoloaded, but in fact, it's there because `loaded` wasn't quite cleaned up properly!

### Expected behavior

`unloadable` constants should play fine with autoloading; they should unload and reload regardless of how they're loaded. 

### Actual behavior

`unloadable` constants which are autoloaded crash on the second reload.

### System configuration

**Rails version**: master 
**Ruby version**: 2.4.1

### Fix? Workaround? 

I think I could modify `AS::D` a bit to support this, is that desirable? 

Or is this contrary to the intended use of `unloadable`?

In this case, we're using `unloadable` as part of github/graphql-client. It's required because `graphql-client` builds a lot of child modules on the fly, based on the contents of view files. So, whenever we reload, we want to reload the _whole_ module, so that its children will be reloaded, too. It's a crazy world, but that's where we are :laughing:

### PR DESCRIPTION
**Pull request by rmosolgo**
Fixes #32359, see that issue for a detailed description.

Is this a desirable solution to that problem? Or is there another approach to try?

### CODE REVIEW

File: activesupport/test/dependencies_test.rb [hunk 1]
```diff
@@ -240,6 +240,46 @@ class AnotherConstant::ReloadError
...
+
+      # Assert that AS::D's internal state matches Ruby's state:
+      assert defined?(UnloadableExample)
+      assert_equal ["UnloadableExample"], ActiveSupport::Dependencies.explicitly_unloadable_constants
+      assert_equal ["UnloadableExample"], ActiveSupport::Dependencies.autoloaded_constants
+      assert_equal 1, ActiveSupport::Dependencies.loaded.size
+
+      # Unload it, and do a reality check that Ruby's state and AS::D's state match
+      ActiveSupport::Dependencies.remove_unloadable_constants!
+      refute defined?(UnloadableExample), "The constant was actually removed"
```
Comment (composerinteralia) [inline activesupport/test/dependencies_test.rb]: Prefer `assert_not` to `refute` (http://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#follow-the-coding-conventions)

File: activesupport/lib/active_support/dependencies.rb [hunk 2]
```diff
@@ -723,6 +723,10 @@ def remove_constant(const) #:nodoc:
       # call from a class or module that were not autoloaded, as in the
       # example above with Object, accessing to that constant must err.
       unless parent.autoload?(to_remove)
+        # Don't try to unload a non-existent constant
+        if !parent.const_defined?(to_remove)
```
Comment (composerinteralia) [inline activesupport/lib/active_support/dependencies.rb]: Can we put this guard clause on one line? `return unless parent.const_defined?(to_remove)`

File: activesupport/test/dependencies_test.rb [hunk 3]
```diff
@@ -240,6 +240,46 @@ class AnotherConstant::ReloadError
...
+
+  # Regression see https://github.com/rails/rails/issues/32359
+  def test_explicit_unloadable_constants_can_are_cleanly_unloaded
+    with_loading "autoloading_fixtures/explicit_unloadable" do
+      # autoload this constant:
+      assert UnloadableExample
+
+      # Assert that AS::D's internal state matches Ruby's state:
+      assert defined?(UnloadableExample)
+      assert_equal ["UnloadableExample"], ActiveSupport::Dependencies.explicitly_unloadable_constants
```
Comment (composerinteralia) [inline activesupport/test/dependencies_test.rb]: Is there a reason you match against the array here, but only check the size on 252?

File: activesupport/test/dependencies_test.rb (hunk 1 above)
Comment (rmosolgo) [reply activesupport/test/dependencies_test.rb]: :+1: Definitely, thanks for the link!

File: activesupport/lib/active_support/dependencies.rb (hunk 2 above)
Comment (rmosolgo) [reply activesupport/lib/active_support/dependencies.rb]: Sure, either way works for me!

File: activesupport/test/dependencies_test.rb (hunk 3 above)
Comment (rmosolgo) [reply activesupport/test/dependencies_test.rb]: No good  reason -- I was just using the test suite as I worked, and a length mismatch wasn't enough information to realize why my test case wasn't failing properly! So I spelled out the whole array here for a more helpful error message when the assertion failed. Happy to standardize these if it's important to do so!

**Comment by rmosolgo** [comment]
I'm going to look more at fixing this in our app.

**Comment by rafaelfranca** [comment]
@rmosolgo I think this behavior is wrong in the framework and your fix make sense to me.

@fxn what do you think?

**Comment by fxn** [comment]
Interesting catch, I'll study it.

**Comment by fxn** [comment]
Hey @rmosolgo.

The plan is to drop this feature, which is your use case?

**Comment by rmosolgo** [comment]
We had bugs with the old autoloader, but we've been on Zeitwerk for a long time. Whatever _used_ to be wrong, it isn't bothering us anymore! 🙇 !

**Pull request closed without merging**